DataParallelTable.lua

```diff
@@ -494,106 +371,318 @@ function DataParallelTable:name()
+-- out if it doesn;t have enough available
+function DataParallelTable:__read(file, version)
+   if version < 2 then
+      local var = file:readObject()
+      for k, v in pairs(var) do
+         self[k] = v
+      end
+      return
+   end
+
+   -- Pre-read gpuAssignments and either use them of ignore them depending on
+   -- whether DataParallelTable.deserializeNGPUs is set.
+   local gpuAssignments = file:readObject()
+   if DataParallelTable.deserializeNGPUs then
+      gpuAssignments = {}
+      for i = 1, DataParallelTable.deserializeNGPUs do gpuAssignments[i] = i end
+      if DataParallelTable.deserializeNGPUs > cutorch.getDeviceCount() then
+         error('Deserialization requested on too many GPUs: ' ..
+                  DataParallelTable.deserializeNGPUs .. ' vs ' ..
+                  cutorch.getDeviceCount() .. ' available')
+      end
+   end
+
+   -- If DataParallelTable.deserializeNGPUs, deserialization overrides
+   -- gpu assignments anyway. If not, we need as many GPUs as the max,
+   -- there may be holes.
+   local nGPUs = math.max(unpack(gpuAssignments))
+   if nGPUs > cutorch.getDeviceCount() then
+      error('Model was serialized on ' ..
+               math.max(unpack(gpuAssignments)) ..
+               ' nGPUs, but you are running on ' .. cutorch.getDeviceCount() ..
+               ' please set DataParallelTable.deserializeNGPUs to ignore ' ..
+               ' serialized tower-GPU assignments')
+   end
+
+   local gpu = cutorch.getDevice()
+   cutorch.setDevice(gpuAssignments[1])
+   -- Deserialize from table
+   local var = file:readObject()
+   for k, v in pairs(var) do
+      self[k] = v
+   end
+
+   if self.usenccl then
+      self.usenccl = pcall(require, 'nccl')
+   end
+   if not self.impl then
+      self.impl = Impls.Basic(self)
+   end
+
+   -- use previously deserialize / recomputed gpuAssignments
+   self.gpuAssignments = gpuAssignments
+   assert(#self.modules == 1)
+
+   local flattenedParams = self.flattenedParams
+   if flattenedParams then
+      self.flattenedParams = self.impl:exec(function(m, i)
+         if i == 1 then
+            return flattenedParams[1]
+         else
+            return { m:getParameters() }
+         end
+      end)
+   end
+end
+
+function DataParallelTable:__write(file)
+   -- Prewrite the current assignments, we may need them to
+   -- deserialize the first tower
+   file:writeObject(self.gpuAssignments)
+   -- Convert to table
+   local t = {}
+   for k, v in pairs(self) do
+      -- Only keep the first tower, this is the way DPT must be used
+      if k  == 'flattenedParams' then
+         t[k] = {v[1]}
+      elseif k == 'inputGpu' or k == 'outputGpu' or k == 'gradInputGpu' or k == 'gradOutputGpu' then
+         t[k] = {}
+      else
+         t[k] = v
+      end
+   end
+   file:writeObject(t)
+   -- Force synchronization, this keeps you honest
+   self:syncParameters()
+end
+
+function DataParallelTable:sanitize()
```

Comment:
This function should be called [clearStates](https://github.com/torch/nn/pull/526)
Also, I think it should return `self`